runtime/lua/vim/fs.lua

```diff
@@ -164,16 +170,18 @@ function M.find(names, opts)
 
       for other, type in M.dir(dir) do
         local f = dir .. '/' .. other
-        for _, name in ipairs(names) do
-          if name == other and (not opts.type or opts.type == type) then
-            if add(f) then
-              return matches
+        if opts.filter == nil or opts.filter(f) then
```

Comment:
What about adding the `filter` check to the `add` function? That way you get it "for free" in both upward and downward searches.

```diff
diff --git a/runtime/lua/vim/fs.lua b/runtime/lua/vim/fs.lua
index 138baa93d..7548aec8f 100644
--- a/runtime/lua/vim/fs.lua
+++ b/runtime/lua/vim/fs.lua
@@ -120,6 +120,10 @@ function M.find(names, opts)
 
   ---@private
   local function add(match)
+    if opts.filter and not opts.filter(match) then
+      return false
+    end
+
     matches[#matches + 1] = match
     if #matches == limit then
       return true
```

Reply:
I guess that would only prevent it from adding it to result, not from searching(?)

Reply:
Yes good point